# Telco Customer Churn Analysis

This project looks at customer churn in a telecom dataset.

The first step is to inspect the raw data and check its structure, missing values, duplicates, data types, and possible outliers before making any changes.

## 1. Setup and Data Loading

The required libraries are imported first, followed by the raw customer dataset.

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("./dataset/telco_customer_churn.csv")
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [3]:
df.shape

(7043, 21)

The dataset contains customer-level records with 7043 rows and 21 columns.

In [4]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [6]:
df.isnull()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7038,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
7039,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
7040,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
7041,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False


In [7]:
df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

The initial missing-value check shows no NaN values. I will also check for blank or whitespace-only values because they may not be detected by isnull().

In [8]:
df.duplicated().sum()

np.int64(0)

In [9]:
(df.astype(str).apply(lambda col: col.str.strip().eq("").sum()))

customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64

There are no duplicate rows in the dataset.

In [10]:
df['gender'].unique()

array(['Female', 'Male'], dtype=object)

In [11]:
df['Partner'].unique()

array(['Yes', 'No'], dtype=object)

In [12]:
df['Dependents'].unique()

array(['No', 'Yes'], dtype=object)

In [13]:
df['Churn'].unique()

array(['No', 'Yes'], dtype=object)

In [14]:
df['Contract'].unique()

array(['Month-to-month', 'One year', 'Two year'], dtype=object)

In [15]:
df['InternetService'].unique()

array(['DSL', 'Fiber optic', 'No'], dtype=object)

In [16]:
df['PaymentMethod'].unique()

array(['Electronic check', 'Mailed check', 'Bank transfer (automatic)',
       'Credit card (automatic)'], dtype=object)

In [17]:
df['MultipleLines'].unique()

array(['No phone service', 'No', 'Yes'], dtype=object)

In [18]:
df['OnlineSecurity'].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [19]:
df['TechSupport'].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

### Checking TotalCharges

TotalCharges contains values that look numeric but is stored as an object column. I will check whether all values can be converted to numeric values.

In [20]:
display(df["TotalCharges"].head())
display(df["TotalCharges"].unique())

0      29.85
1     1889.5
2     108.15
3    1840.75
4     151.65
Name: TotalCharges, dtype: object

array(['29.85', '1889.5', '108.15', ..., '346.45', '306.6', '6844.5'],
      shape=(6531,), dtype=object)

In [21]:
total_charges_numeric = pd.to_numeric(df["TotalCharges"], errors="coerce")

total_charges_numeric.isnull().sum()

np.int64(11)

In [22]:
df.loc[total_charges_numeric.isnull(), ["customerID", "tenure", "TotalCharges"]]

,customerID,tenure,TotalCharges
488,4472-LVYGI,0,
753,3115-CZMZD,0,
936,5709-LVOEQ,0,
1082,4367-NUYAO,0,
1340,1371-DWPAZ,0,
3331,7644-OMVMY,0,
3826,3213-VVOLG,0,
4380,2520-SGTTA,0,
5218,2923-ARZLG,0,
6670,4075-WKNIU,0,


The TotalCharges column contains blank values that were not detected by the initial isnull() check because they were stored as strings. These values occur for customers with zero tenure.

In [23]:
df.describe()

,SeniorCitizen,tenure,MonthlyCharges
count,7043.000000,7043.000000,7043.000000
mean,0.162147,32.371149,64.761692
std,0.368612,24.559481,30.090047
min,0.000000,0.000000,18.250000
25%,0.000000,9.000000,35.500000
50%,0.000000,29.000000,70.350000
75%,0.000000,55.000000,89.850000
max,1.000000,72.000000,118.750000


In [24]:
df.describe(include="object")

,customerID,gender,Partner,Dependents,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,TotalCharges,Churn
count,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043
unique,7043,2,2,2,2,3,3,3,3,3,3,3,3,3,2,4,6531,2
top,7590-VHVEG,Male,No,No,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,20.2,No
freq,1,3555,3641,4933,6361,3390,3096,3498,3088,3095,3473,2810,2785,3875,4171,2365,11,5174


### Checking for numerical outliers

I will use the IQR method to identify unusually low or high values in the numerical columns.

In [25]:
numeric_columns = ["tenure", "MonthlyCharges"]

for column in numeric_columns:
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    
    outliers = df[(df[column] < lower_bound) | (df[column] > upper_bound)]
    
    print(f"{column}:")
    print(f"Lower bound: {lower_bound:.2f}")
    print(f"Upper bound: {upper_bound:.2f}")
    print(f"Number of outliers: {len(outliers)}")
    print()

tenure:
Lower bound: -60.00
Upper bound: 124.00
Number of outliers: 0

MonthlyCharges:
Lower bound: -46.02
Upper bound: 171.38
Number of outliers: 0



The IQR method identifies values that are outside the usual range. These values will not be removed at this stage because they may represent genuine customers rather than data errors.

## Audit Summary

The dataset contains 7,043 customer records and 21 columns.

The initial missing-value check showed no standard NaN values. A separate check for blank values was also performed to identify values that may not be recognised by isnull().

The duplicate check was performed using the full dataset.

The main data type issue found was TotalCharges, which is stored as an object even though it contains numeric-looking values. This column was checked using numeric conversion to identify values that could not be converted.

The numerical columns were also checked for potential outliers using the IQR method. The identified values were kept in the raw dataset because an unusual value does not automatically mean that the record is incorrect.

No cleaning or feature engineering was performed during this audit stage.

## 2. Data Cleaning and Preprocessing

The raw data has been audited, so the next step is to fix the issues found during the audit and prepare the data for the later analysis and modelling stages.

In [26]:
df_clean = df.copy()

In [27]:
df_clean.shape

(7043, 21)

In [28]:
df_clean["TotalCharges"] = pd.to_numeric(df_clean["TotalCharges"], errors="coerce")

In [29]:
df_clean["TotalCharges"]

0         29.85
1       1889.50
2        108.15
3       1840.75
4        151.65
         ...   
7038    1990.50
7039    7362.90
7040     346.45
7041     306.60
7042    6844.50
Name: TotalCharges, Length: 7043, dtype: float64

In [30]:
df_clean["TotalCharges"].dtype

dtype('float64')

In [31]:
df_clean["TotalCharges"].isnull().sum()

np.int64(11)

In [32]:
df_clean[df_clean["TotalCharges"].isnull()][
    ["customerID", "tenure", "TotalCharges"]
]

,customerID,tenure,TotalCharges
488,4472-LVYGI,0,NaN
753,3115-CZMZD,0,NaN
936,5709-LVOEQ,0,NaN
1082,4367-NUYAO,0,NaN
1340,1371-DWPAZ,0,NaN
3331,7644-OMVMY,0,NaN
3826,3213-VVOLG,0,NaN
4380,2520-SGTTA,0,NaN
5218,2923-ARZLG,0,NaN
6670,4075-WKNIU,0,NaN


In [33]:
df_clean.loc[df_clean["TotalCharges"].isnull(), "tenure"].value_counts()

tenure
0    11
Name: count, dtype: int64

### TotalCharges

The `TotalCharges` column was originally stored as an object even though it represents a numerical value. After converting it to numeric, the blank values were represented as missing values.

The affected records were checked against `tenure` to understand why the values were missing before deciding how to handle them.

### 2.1 Handling missing TotalCharges

In [34]:
df_clean["TotalCharges"] = df_clean["TotalCharges"].fillna(0)

In [35]:
df_clean["TotalCharges"].isnull().sum()

np.int64(0)

### 2.2 Checking missing values

In [36]:
df_clean.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

### 2.3 Separating features and target

In [37]:
X = df_clean.drop(columns=["Churn", "customerID"])
y = df_clean["Churn"]

In [38]:
X.shape

(7043, 19)

In [39]:
y.shape

(7043,)

In [40]:
y.value_counts()

Churn
No     5174
Yes    1869
Name: count, dtype: int64

In [41]:
y.value_counts(normalize=True)

Churn
No     0.73463
Yes    0.26537
Name: proportion, dtype: float64

In [42]:
y = y.map({"Yes": 1, "No": 0})

In [43]:
y.value_counts()

Churn
0    5174
1    1869
Name: count, dtype: int64

### 2.4 Train/test split

In [44]:
from sklearn.model_selection import train_test_split

In [45]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [46]:
X_train.shape

(5634, 19)

In [47]:
X_test.shape

(1409, 19)

In [48]:
y_train.value_counts(normalize=True)

Churn
0    0.734647
1    0.265353
Name: proportion, dtype: float64

In [49]:
y_test.value_counts(normalize=True)

Churn
0    0.734564
1    0.265436
Name: proportion, dtype: float64

### 2.5 Identifying numerical and categorical columns

In [51]:
numerical_features = X_train.select_dtypes(include=["int64", "float64"]).columns

In [52]:
categorical_features = X_train.select_dtypes(include=["object"]).columns

In [53]:
display(numerical_features)
display(categorical_features)

Index(['SeniorCitizen', 'tenure', 'MonthlyCharges', 'TotalCharges'], dtype='object')

Index(['gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines',
       'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
       'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract',
       'PaperlessBilling', 'PaymentMethod'],
      dtype='object')

### 2.6 Building the preprocessing steps

In [54]:
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

In [55]:
numerical_pipeline = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", StandardScaler())
])

In [56]:
categorical_pipeline = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("encoder", OneHotEncoder(handle_unknown="ignore"))
])

In [57]:
preprocessor = ColumnTransformer(
    transformers=[
        ("num", numerical_pipeline, numerical_features),
        ("cat", categorical_pipeline, categorical_features)
    ]
)

### 2.7 Fitting preprocessing on the training data

In [58]:
X_train_processed = preprocessor.fit_transform(X_train)

In [59]:
X_test_processed = preprocessor.transform(X_test)

In [61]:
X_train_processed.shape

(5634, 45)

In [62]:
X_test_processed.shape

(1409, 45)

In [63]:
y_train.shape

(5634,)

In [64]:
y_test.shape

(1409,)

### 2.8 Preprocessing summary

The cleaned dataset was split into training and test sets before preprocessing. Numerical variables were handled using median imputation and standard scaling, while categorical variables were handled using most-frequent imputation and one-hot encoding. The preprocessing steps were fitted only on the training data and then applied to the test data to avoid data leakage.

### Avoiding data leakage

The train/test split was performed before fitting the preprocessing transformers. Imputation, encoding, and scaling were learned from the training data only. The same fitted transformations were then applied to the test data.